craft_providers/actions/snap_installer.py

```diff
@@ -0,0 +1,191 @@
+    ]
+
+    logger.debug("Executing command on host: %s", shlex.join(cmd))
+    subprocess.run(
+        cmd,
+        capture_output=True,
+        check=True,
+    )
+
+
+@contextlib.contextmanager
+def _get_host_snap(snap_name: str) -> Iterator[pathlib.Path]:
+    """Get snap installed on host.
+
+    Snapd provides an API to fetch a snap. First use that to fetch a snap.
+    If the snap is installed using `snap try`, it may fail to download. In
+    that case, attempt to construct the snap by packing it ourselves.
+
+    :yields: Path to snap which will be cleaned up afterwards.
+    """
+    with temp_paths.home_temporary_directory() as tmp_dir:
+        snap_path = tmp_dir / f"{snap_name}.snap"
+
+        try:
+            _download_host_snap(snap_name=snap_name, output=snap_path)
+        except SnapInstallationError:
+            logger.warning(
+                "Failed to fetch snap from snapd, falling back to `snap pack` to recreate."
+            )
+            _pack_host_snap(snap_name=snap_name, output=snap_path)
+
+        yield snap_path
+
+
+def inject_from_host(*, executor: Executor, snap_name: str, classic: bool) -> None:
+    """Inject snap from host snap.
+
+    :raises SnapInstallationError: on unexpected error.
+    """
+    target_snap_path = pathlib.Path(f"/tmp/{snap_name}.snap")
+
+    try:
+        # Clean outdated snap, if exists.
+        executor.execute_run(
+            ["rm", "-f", target_snap_path.as_posix()],
+            check=True,
+            capture_output=True,
+        )
+
+        with _get_host_snap(snap_name) as host_snap_path:
+            executor.push_file(
+                source=host_snap_path,
+                destination=target_snap_path,
+            )
+
+        executor.execute_run(
+            snap_cmd.formulate_install_command(
+                classic=classic, dangerous=True, snap_path=target_snap_path
+            ),
+            check=True,
+            capture_output=True,
+        )
+    except subprocess.CalledProcessError as error:
+        raise SnapInstallationError(
+            brief=f"Failed to inject snap {snap_name!r}.",
+            details=details_from_called_process_error(error),
+        ) from error
+    except ProviderError as error:
+        raise SnapInstallationError(
+            brief=f"Failed to inject snap {snap_name!r}.",
+        ) from error
+
+
+def install_from_store(
+    *,
+    executor: Executor,
+    snap_name: str,
+    channel: str,
+    classic: bool,
+) -> None:
+    """Install snap from store into target.
+
+    Perform installation using method which prevents refreshing.
+
+    :param executor: Executor for target.
+    :param snap_name: Name of snap to install.
+    :param channel: Channel to install from.
+    :param classic: Install in classic mode.
+
+    :raises SnapInstallationError: on unexpected error.
+    """
+    target_snap_path = pathlib.Path(f"/tmp/{snap_name}.snap")
+
+    try:
+        executor.execute_run(
```

Comment:
I believe "snap download" caches and knows how to resume, why not download on host and push to avoid downloading the same thing potentially N times? Auto prunning should be fine.

Reply:
I was looking ahead to cross-platform support where the host would not support snaps.

Generally, if the host supports snaps, the application is likely to be injected rather than installed from store.

Reply:
snap download should work on osx and windows.

Reply:
Alright, if there is no issue with making snap a dependency, I update to use the host np